param_passing/proposal.md

```diff
@@ -0,0 +1,465 @@
+    <camera name="cam">
+      ...
+    </camera>
+  </sensor>
+</link>
+```
+
+If the user wants to specify the `//camera` element, then
+`element_id="chassis::camera::cam"` where "chassis" is the name of the `//link`,
+"camera" is the name of the `//link/sensor`, and "cam" is the name of the
+`//link/sensor/camera`. This is how the correct element will be identified and
+will be called the element identifier. Although it is possible to identify the
+element without requiring the user to specify the tag, enforcing the user to
+provide the tag allows less ambiguity to the user as well as provides a means of
+error checking. Then the user will need to specify a corresponding action with
+that element to dictate the alteration they would like to make using a new
+attribute `action`. The attribute takes in a string and the available actions
+are:
+
+* `add`: adds new elements to the original model
+* `modify`: modifies values and/or attributes of elements.
+This only updates existing elements and does not add or remove them.
+* `remove`: removes the elements from the original model
+* `replace`: replaces the elements from the original model to the new provided elements
+
+The `action` attribute must be provided either in the element identifier or in
+the direct children of the identifier.
+
+##### Alternatives considered
+
+Alternatively, instead of using the proposed element identifier, the user could
+use the attribute `id` for each customizable element in the original file but
+the intent of using the named path (i.e., the name of the parent elements and
+itself separated by double colons) was to provide a way for the user to modify
+the original model without requiring large amounts of updates to the original
+file.
+
+## Examples
+
+### Example 1
+
+Let's look at an example, here is an original model `base_robot`:
+
+```xml
+<model name="base_robot">
+  ...
+  <link name="chassis">
+    <visual name="lidar_visual">
+      ...
+    </visual>
+    <sensor name="lidar">
+      ...
+    </sensor>
+    <sensor name="camera">
+      <pose>0.5 0.02 0 0 0</pose>
+      ...
+    </sensor>
+  </link>
+
+  <link name="top">
+    <visual name="camera_visual">
+      ...
+      <geometry>
+        <box>
+          <size>0.03 0.03 0.05</size>
+        </box>
+      </geometry>
+      <material>
+        <ambient>0.5 0.5 0 1</ambient>
+        <diffuse>0 1 0 1</diffuse>
+        <specular>0.5 0.5 0.5 1</specular>
+      </material>
+    </visual>
+    <sensor name="camera">
+      <pose>0.2 1 0.5 0 0 0</pose>
+      <update_rate>20</update_rate>
+      ...
+    </sensor>
+  </link>
+```
+
+Now the user would like to use the original model `base_robot` in a world file
+but with modifications. The modifications include removing the lidar from the
+chassis link, adding a plugin to the camera sensor in the chassis link, adding a
+camera visual to the chassis link, update camera information in the top link,
+and replace the top link's camera visual geometry.
```

Comment:
also updating camera visual material?

Maybe itemize these actions using bullet points.

Reply:
How's this? d6aa791a54a569dc01bb88789b95a35a0f14fac2